File: bms/go/geo/geo_r3_Vector_Normalize/geo_r3_Vector_Normalize.go
package test

import (
	"math"
)

func (v geo_r3_Vector) Normalize() geo_r3_Vector {
	n2 := v.Norm2()
	if n2 == 0 {
		return geo_r3_Vector{0, 0, 0}
	}
	return v.Mul(1 / math.Sqrt(n2))
}

func (v geo_r3_Vector) Norm2() float64 { return v.Dot(v) }

func (v geo_r3_Vector) Dot(ov geo_r3_Vector) float64 {
	return float64(v.X*ov.X) + float64(v.Y*ov.Y) + float64(v.Z*ov.Z)
}

func (v geo_r3_Vector) Mul(m float64) geo_r3_Vector { return geo_r3_Vector{m * v.X, m * v.Y, m * v.Z} }

type geo_r3_Vector struct {
	X, Y, Z float64
}

File: translations_only/success/go/geo/geo_r3_Vector_Normalize/gpt4/CAPR/geo_r3_Vector_Normalize.rs

struct GeoR3Vector {
    x: f64,
    y: f64,
    z: f64,
}

fn normalize(v: &GeoR3Vector) -> GeoR3Vector {
    let n2 = norm2(v);
    if n2 == 0.0 {
        GeoR3Vector { x: 0.0, y: 0.0, z: 0.0 }
    } else {
        mul(v, 1.0 / n2.sqrt())
    }
}

fn norm2(v: &GeoR3Vector) -> f64 {
    dot(v, v)
}

fn dot(v: &GeoR3Vector, ov: &GeoR3Vector) -> f64 {
    (v.x * ov.x) + (v.y * ov.y) + (v.z * ov.z)
}

fn mul(v: &GeoR3Vector, m: f64) -> GeoR3Vector {
    GeoR3Vector { x: m * v.x, y: m * v.y, z: m * v.z }
}
